### PR TITLE
Updated snapshot docs with ARM kernel parameter induced latency

### PR DESCRIPTION
**Pull request by georgepisaltu**
# Reason for This PR
Running with the host kernel command line parameter `console=AMA0` on `aarch64` was degrading the snapshot restore latency from 3ms to 8.5ms.

## Description of Changes

Added an entry in the snapshot performance known issues to highlight this behavior.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: docs/snapshotting/snapshot-support.md [hunk 1]
```diff
@@ -123,6 +123,9 @@ The baseline for snapshot resume latency target on Intel is under **8ms** with
   deal with cryptographic secrets. Please see [Snapshot security and uniqueness](#snapshot-security-and-uniqueness).
 - Snapshotting on arm64 works for both GICv2 and GICv3 enabled guests.
   However, restoring between different GIC version is not possible.
+- Running with the host kernel command line parameter `console=ttyAMA0` on
+  `aarch64` degrades the snapshot restore time to 8.5ms because the host kernel
+  logs generated in the restore process are slower with this configuration.
```
Comment (serban300) [inline docs/snapshotting/snapshot-support.md:128]: This configuration leads to a degradation in performance in multiple scenarios, not only on snapshot restore. Would it be better to mention it in the `host-prod-setup` ?

Comment (alindima) [reply docs/snapshotting/snapshot-support.md:128]: +1, I think it is correlated to the entry we have in the prod-host-setup on the serial device

Comment (gc-plp) [reply docs/snapshotting/snapshot-support.md:128]: The statement is not completely correct as `ttyAMA0` is the name given to the serial console for specific ARM boards that have a specific serial controller, as seen [here](https://github.com/torvalds/linux/blob/v4.16/drivers/tty/serial/amba-pl011.c#L2488). Coincidence makes it that the controller that determines the name is quite popular and most ARM boards will have a `ttyAMA0` port.

A better way of explaining the problem would be to explain the actual effect and mitigation. I'm not 100% sure that this doesn't happen on x86 when enabling console output on `ttyS0`.

As @serban300 mentioned, I'd put this under the prod host setup chapter.

**Review by georgepisaltu — COMMENTED** [review]
Updated `prod-host-setup.md` as per suggestions. @serban300 @alindima @gc-plp PTAL!

File: docs/prod-host-setup.md [hunk 2]
```diff
@@ -46,6 +46,21 @@ for consuming and storing this data safely. We suggest using any upper-bounded
 forms of storage, such as fixed-size or ring buffers, programs like `journald`
 or `logrotate`, or redirecting to a named pipe.
 
+### Logging and performance
+
+There is extra latency introduced in Firecracker functionality as the process
+will log various error or warning messages. Some host configurations can have
+an effect on the performance of the process as Firecracker still has to use
+the host devices, such as the serial device, to write the logs generated
```
Comment (serban300) [inline docs/prod-host-setup.md:54]: I think this should be rephrased. Firecracker doesn't use the serial device. Host kernel logs are generated during normal Firecracker operations. And these logs can end up on the serial console depending on the host config.

**Review by gc-plp — COMMENTED** [review]
>We recommend that users disable logging to the host serial console entirely and alternatively use a faster methods to get the logs, as described in the section above.

Kernel logs and firecracker logs are different. We do not actually recommend getting the kernel logs in the section above.

I'd rephrase this as:

> We recommend adding `quiet loglevel=1` to the kernel command line to limit the number of messages outputted on the serial console.

**Review by georgepisaltu — COMMENTED** [review]
@serban300 @gc-plp @alindima PTAL!

File: docs/prod-host-setup.md [hunk 3]
```diff
@@ -46,6 +46,20 @@ for consuming and storing this data safely. We suggest using any upper-bounded
 forms of storage, such as fixed-size or ring buffers, programs like `journald`
 or `logrotate`, or redirecting to a named pipe.
 
+### Logging and performance
+
+There is extra latency introduced in Firecracker functionality as the process
+generates and logs various error or warning messages. Some host configurations
```
Comment (serban300) [inline docs/prod-host-setup.md:52]: I think we can delete the first phrase.

File: docs/prod-host-setup.md [hunk 4]
```diff
@@ -46,6 +46,20 @@ for consuming and storing this data safely. We suggest using any upper-bounded
...
+can have an effect on the performance of the process as Firecracker will
+generate host kernel logs during normal operations. The most recent example of
+this was the addition of `console=ttyAMA0` kernel command line argument on one
```
Comment (serban300) [inline docs/prod-host-setup.md:55]: I would specify clearly that we're talking about the host kernel.

File: docs/prod-host-setup.md [hunk 5]
```diff
@@ -46,6 +46,20 @@ for consuming and storing this data safely. We suggest using any upper-bounded
 forms of storage, such as fixed-size or ring buffers, programs like `journald`
 or `logrotate`, or redirecting to a named pipe.
 
+### Logging and performance
+
+There is extra latency introduced in Firecracker functionality as the process
+generates and logs various error or warning messages. Some host configurations
+can have an effect on the performance of the process as Firecracker will
+generate host kernel logs during normal operations. The most recent example of
```
Comment (serban300) [inline docs/prod-host-setup.md:54]: I would separate the example in a separate paragraph.

File: docs/prod-host-setup.md (hunk 3 above)
Comment (georgepisaltu) [reply docs/prod-host-setup.md:52]: Done.

File: docs/prod-host-setup.md (hunk 5 above)
Comment (georgepisaltu) [reply docs/prod-host-setup.md:54]: Done.

File: docs/prod-host-setup.md (hunk 4 above)
Comment (georgepisaltu) [reply docs/prod-host-setup.md:55]: Done.

File: docs/prod-host-setup.md [hunk 6]
```diff
@@ -46,6 +46,20 @@ for consuming and storing this data safely. We suggest using any upper-bounded
...
+as Firecracker will generate host kernel logs during normal operations.
+
+The most recent example of this was the addition of `console=ttyAMA0` host
+kernel command line argument on one of our testing setups which, because this
+particular console made writing the logs slower than the previous one,
+degraded the snapshot restore time from 3ms to 8.5ms on `aarch64`. This
+happened as creating the tap device for snapshot restore generated host kernel
+logs, which were very slow to write. We recommend adding `quiet loglevel=1` to
+the host kernel command line to limit the number of messages written to the
+serial console.
```
Comment (serban300) [inline docs/prod-host-setup.md:61]: This last phrase seems to belong inside the first paragraph.

File: docs/prod-host-setup.md [hunk 7]
```diff
@@ -46,6 +46,20 @@ for consuming and storing this data safely. We suggest using any upper-bounded
...
 
+### Logging and performance
+
+Some host configurations can have an effect on the performance of the process
+as Firecracker will generate host kernel logs during normal operations.
+
+The most recent example of this was the addition of `console=ttyAMA0` host
+kernel command line argument on one of our testing setups which, because this
+particular console made writing the logs slower than the previous one,
+degraded the snapshot restore time from 3ms to 8.5ms on `aarch64`. This
```
Comment (serban300) [inline docs/prod-host-setup.md:57]: I would simplify this a bit. Something like: `The most recent example of this was the addition of `console=ttyAMA0` host kernel command line argument on one of our testing setups. This enabled console logging, which degraded the snapshot restore time from 3ms to 8.5ms on aarch64.`

File: docs/prod-host-setup.md (hunk 6 above)
Comment (georgepisaltu) [reply docs/prod-host-setup.md]: Done/

File: docs/prod-host-setup.md (hunk 7 above)
Comment (georgepisaltu) [reply docs/prod-host-setup.md]: Done.

**Review by serban300 — APPROVED** [review]

**Review by alindima — APPROVED** [review]
